redis/src/connection.rs

```diff
@@ -1667,39 +1667,61 @@ impl<'a> PubSub<'a> {
         }
     }
 
-    fn cache_messages_until_received_response(&mut self, cmd: &mut Cmd) -> RedisResult<()> {
-        if self.con.protocol != ProtocolVersion::RESP2 {
-            cmd.set_no_response(true);
-        }
-        let mut response = cmd.query(self.con)?;
+    fn cache_messages_until_received_response(
+        &mut self,
+        cmd: &mut Cmd,
+        is_sub_unsub: bool,
+    ) -> RedisResult<Value> {
+        let ignore_response = self.con.protocol != ProtocolVersion::RESP2 && is_sub_unsub;
+        cmd.set_no_response(ignore_response);
+
+        self.con.send_packed_command(&cmd.get_packed_command())?;
+
         loop {
-            if let Some(msg) = Msg::from_owned_value(response) {
+            let response = self.con.recv_response()?;
+            if let Some(msg) = Msg::from_value(&response) {
                 self.waiting_messages.push_back(msg);
             } else {
-                return Ok(());
+                return Ok(response);
             }
-            response = self.con.recv_response()?;
         }
     }
 
     /// Subscribes to a new channel.
     pub fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        self.cache_messages_until_received_response(cmd("SUBSCRIBE").arg(channel))
+        self.cache_messages_until_received_response(cmd("SUBSCRIBE").arg(channel), true)?;
```

Comment:
Why all pub/sub are returning Ok and not the response? It is valuable information.

Reply:
That's the current behavior, which I'm not changing. Changing the return type of the function is a breaking change.